src/transformers/models/instructblip/modeling_instructblip.py

```diff
@@ -0,0 +1,1968 @@
+
+            if attention_mask is None:
+                attention_mask = torch.ones_like(input_ids)
+            expected_device = language_model_attention_mask.device
+            attention_mask = torch.cat([language_model_attention_mask, attention_mask.to(expected_device)], dim=1)
+
+            if self.config.use_decoder_only_language_model:
+                outputs = self.language_model(
+                    inputs_embeds=inputs_embeds,
+                    attention_mask=attention_mask,
+                    output_attentions=output_attentions,
+                    output_hidden_states=output_hidden_states,
+                    return_dict=return_dict,
+                )
+                logits = outputs.logits if return_dict else outputs[0]
+                loss = None
+                # we compute the loss here since we need to take into account the sequence length of the query embeds
+                if labels is not None:
+                    labels = labels.to(logits.device)
+                    logits = logits[:, -labels.size(1) :, :]
+                    # Shift so that tokens < n predict n
+                    shift_logits = logits[..., :-1, :].contiguous()
+                    shift_labels = labels[..., 1:].contiguous().to(logits.device)
+
+                    # Flatten the tokens
+                    loss_fct = CrossEntropyLoss(reduction="mean")
+
+                    loss = loss_fct(shift_logits.view(-1, self.config.text_config.vocab_size), shift_labels.view(-1))
+            else:
+                outputs = self.language_model(
+                    inputs_embeds=inputs_embeds,
+                    attention_mask=attention_mask,
+                    decoder_input_ids=decoder_input_ids,
+                    decoder_attention_mask=decoder_attention_mask,
+                    output_attentions=output_attentions,
+                    output_hidden_states=output_hidden_states,
+                    return_dict=return_dict,
+                    labels=labels,
+                )
+                loss = outputs.loss if return_dict else outputs[0]
+                logits = outputs.logits if return_dict else outputs[1]
+
+        if not return_dict:
+            output = (logits, vision_outputs, query_outputs, outputs)
+            return ((loss,) + output) if loss is not None else output
+
+        return InstructBlipForConditionalGenerationModelOutput(
+            loss=loss,
+            logits=logits,
+            vision_outputs=vision_outputs,
+            qformer_outputs=query_outputs,
+            language_model_outputs=outputs,
+        )
+
+    def maybe_autocast(self, dtype=torch.float16):
+        # if on cpu, don't use autocast
+        # if on gpu, use autocast with dtype if provided, otherwise use torch.float16
+        enable_autocast = self.device != torch.device("cpu")
+
+        if enable_autocast:
+            return torch.cuda.amp.autocast(dtype=dtype)
+        else:
+            return contextlib.nullcontext()
```

Comment:
Hmm yes I was afraid this can't be included. However, by not including this, generation performance is significantly worse than the original implementation in LAVIS. Is there an alternative way of approaching this?

I've seen that support for certain models on different hardware (like Habana Gaudi Processors or HPUs) is supported in Optimum, e.g. implementations of GPT-2, BLOOM: https://github.com/huggingface/optimum-habana/tree/main/optimum/habana/transformers/models. Could we have a similar use case here? I feel like TPUs and XPUs represent < 1% of the users

Reply:
Ok just ran some various settings. 

The [original model](https://github.com/salesforce/LAVIS/tree/main/lavis/models/blip2_models) is always using autocasting (float16 for vision encoder, float16 or bfloat16 for language model depending on Vicuna vs Flan-T5). The differences seem to occur only when using beam search:

### Greedy decoding (float32)
Settings for both models:
```
do_sample=False,
num_beams=1,
max_length=256,
min_length=1,
top_p=0.9,
repetition_penalty=1.5,
length_penalty=1.0,
temperature=1,
```

When using float32 for the HF model, greedy decoding gives exactly the same results as the original implementation for both Flan-T5 and Vicuna-7b:

#### Vicuna 7b:

Original generation: ['The unusual aspect of this image is that a man in yellow shirt and shorts, who appears to be ironing clothes on the back seat or roof rack area of his car while driving down an urban street. This situation could potentially pose safety risks for both him as well as other road users due to distractions caused by handling laundry during travel time instead focusing solely on safe vehicle operation responsibilities like traffic laws compliance']

HF generation: ['The unusual aspect of this image is that a man in yellow shirt and shorts, who appears to be ironing clothes on the back seat or roof rack area of his car while driving down an urban street. This situation could potentially pose safety risks for both him as well as other road users due to distractions caused by handling laundry during travel time instead focusing solely on safe vehicle operation responsibilities like traffic laws compliance']

#### Flan-T5-xl:

Original generation: ['The image shows a man ironing clothes on the back of a yellow car in a busy city street. The unusual aspect of the image is that the man is not wearing any clothing, which may indicate that he is working on his own laundry. The image also shows a person walking in the background, which may suggest that the man is taking a break from his work.']

HF generation: ['The image shows a man ironing clothes on the back of a yellow car in a busy city street. The unusual aspect of the image is that the man is not wearing any clothing, which may indicate that he is working on his own laundry. The image also shows a person walking in the background, which may suggest that the man is taking a break from his work.']

### Beam search (both float32)

Settings for both models:

```
do_sample=False,
num_beams=5,
max_length=256,
min_length=1,
top_p=0.9,
repetition_penalty=1.5,
length_penalty=1.0,
temperature=1,
```
For num_beams=5, performance seems worse for the HF Flan-T5 model when using float32. The Vicuna model gives the same results:

#### Vicuna-7b:

Original generation: ['The unusual aspect of this image is that a man is ironing clothes on the back of a yellow SUV, which is parked in the middle of a busy city street. This is an unconventional approach to ironing clothes, as it requires the man to balance himself and his ironing equipment on top of the vehicle while navigating through traffic. Additionally, the presence of taxis and other vehicles in the scene further emphasizes the unusual nature of this situation.']

HF generation: ['The unusual aspect of this image is that a man is ironing clothes on the back of a yellow SUV, which is parked in the middle of a busy city street. This is an unconventional approach to ironing clothes, as it requires the man to balance himself and his ironing equipment on top of the vehicle while navigating through traffic. Additionally, the presence of taxis and other vehicles in the scene further emphasizes the unusual nature of this situation.']

#### Flan-T5-xl:

Original generation: ['The image depicts a man ironing clothes on the back of a yellow car in the middle of a busy city street. The man is wearing a yellow shirt with a yellow tie, and he is using an ironing board attached to the back of the car. The image is unusual in that it shows a man ironing clothes on the back of a car in the middle of a busy city street. The man is using an ironing board attached to the back of a yellow car in the middle of a busy city street. The man is using an ironing board attached to the back of a yellow car in the middle of a busy city street. The image is unusual in that it shows a man ironing clothes on the back of a car in the middle of a busy city street.']

HF generation: ['The image depicts a man ironing clothes on the back of a yellow car in the middle of a busy city street. The man is wearing a yellow shirt with an ironing board attached to it, suggesting that he is using the vehicle as a laundry room. The image is unusual due to the presence of a man in a yellow shirt ironing clothes on the back of a yellow car in the middle of a busy city street.']

### Beam search (using maybe_autocast)

Settings for both models:

```
do_sample=False,
num_beams=5,
max_length=256,
min_length=1,
top_p=0.9,
repetition_penalty=1.5,
length_penalty=1.0,
temperature=1,
```

When casting the vision encoder to float16 and the language model to either float16 or bfloat16 as done in the original implementation, all results are the same:

#### Vicuna-7b:

Original generation: ['The unusual aspect of this image is that a man in yellow shirt and shorts, who appears to be ironing clothes on the back seat or roof rack area of his car while driving down an urban street. This situation could potentially pose safety risks for both him as well as other road users due to distractions caused by handling laundry during travel time instead focusing solely on safe vehicle operation responsibilities like traffic laws compliance']

HF generation: ['The unusual aspect of this image is that a man in yellow shirt and shorts, who appears to be ironing clothes on the back seat or roof rack area of his car while driving down an urban street. This situation could potentially pose safety risks for both him as well as other road users due to distractions caused by handling laundry during travel time instead focusing solely on safe vehicle operation responsibilities like traffic laws compliance']


#### Flan-T5-xl:

Original generation: ['The image shows a man ironing clothes on the back of a yellow car in a busy city street. The unusual aspect of the image is that the man is not wearing any clothing, which may indicate that he is working on his own laundry. The image also shows a person walking in the background, which may suggest that the man is taking a break from his work.']

HF generation: ['The image shows a man ironing clothes on the back of a yellow car in a busy city street. The unusual aspect of the image is that the man is not wearing any clothing, which may indicate that he is working on his own laundry. The image also shows a person walking in the background, which may suggest that the man is taking a break from his work.']

So it seems there's only an issue with the Flan-T5 model in beam search setting if we'd remove autocasting.

Reply:
I'll ask the Salesforce author whether it's fine to merge without using autocasting.